### PR TITLE
lsm303dlhc migration to HAL

### LINKED ISSUE #137: Migrate lsm303dlhc to HAL

### PR DESCRIPTION
**Pull request by WilliamLim781**
TL;DR : added HAL support for lsm303dlhc some tweaks to the init function and compute acceleration

### Summary
used the i2c1 bus from the stm32f429zit to establish an i2c bus with the lsm303dlhc accelerometer. This was done using both the .IOC file to establish i2c-specific parameters. Some of the important ones are in the picture below.
i2c config

copied over the abstract interface and the lsm303dlhc library and removed all Mbed OS I2C Class usages. Most of the work will be in the .cpp file for the lsm303dlhc library as this folder will handle i2c initialization, i2c reading of registers, and i2c writing of registers. in the snippet of code just below here you can see the Mbed version is using a write and read function  that comes with their i2c class.

mbed version

These I2C functions are not supported so these read and write functions need to be replaced with stm32 HAL functions. so instead you need to replace the I2C functions with HAL read and write functions which look like this

stm32cube

Other than the HAL conversion for the i2c bus initialization, read, and write some tweaks were made onto the library. 
   1.) The Mbed library we used had the data as 12-bit data but I couldn't find anywhere that said the lsm303dlhc outputted 12- 
         bit data so I changed it to 16-bit data

   2.) changed ComputeAcceration() function. The data that was being outputted when the accelerometer was stationary was 
        around 16 on the z-axis. To me this didn't make any sense so instead I linearized the accelerometer data based on 
        theoretical data. The accelerometer has a default range of ± 2 g. and the accelerometer will output signed data as the 
        acceleration will have a direction. This meant that the accelerometer has a range of -32768 to 32767 bits ( remember you  
        need 1 bit for 0). so I paired 2g with 32767, -2 g with -32768, 1g with 16348,  -1g -16348, and 0 g with 0. I then graphed all 
        of them using Excel and used the linear regression function to create a linear equation that can take in the bits from the 
        raw accelerometer data and output the g values in m/s^2 ( 9.8 = 1 g etc)

   3.) the .IOC handles initialization of the accelerometer frequency and pins so that was removed from the constructor. In 
        addition, I added the I added the I2C_HandleTypeDef& as a parameter for the constructor to do some dependency 
        injection.

   4.) in the init() function, I changed the value written to the CTRL_REG1_A register from 0x2F to 0x57. The reason for this is 
        that I wanted to set the accelerometer to 100 hz and wanted to disable the low-power mode. The register table for 
        CTRL_REG1_A can be seen below.
table
        I also added 0x08 to  CTRL_REG4_A to enable high-resolution 

#### HAL_MAX_DELAY
when using the HAL transmit and receive function, I use HAL_MAX_DELAY. Which in theory shouldn't be a big deal. but in practice, this could be a problem. The reason is that both the transmit function and the receive function are blocking functions so they will block any processes until they finish. If for whatever reason these functions cannot finish then the code can never progress since HAL_MAX_DELAY is a very very long wait time for the blocking function. A potential problem could be that if the connection on the SDA or SCL line are weak they could potentially fail to send data. If the HAL functions have no way to to detect this then the code will be stuck on the transmit function. Basically, I'm using HAL_MAX_DELAY as a timeout timer. Realistically this shouldn't be a problem but it is something to be cognizant about as it could potentially be a problem in the future.

#### linearization of the accelerometer
under my quick research, this is a valid approach since the accelerometer is supposed to be linear. However, I need further confirmation if this is correct. In addition, I used theoretical values to linearize the accelerometer which may cause a problem. I noticed specifically that the x-axis has had some accuracy issue which can be seen in the picture below when having the x-axis take the acceleration due to gravity.
x
 you can see that the x-axis isn't fully reading 9.8. Obviously, there are variables that may not be fully controller like whether or not the accelerometer's x-axis is completely parallel to the force of gravity. But despite that the reading is still pretty different from 9.8. The other axes are almost fine they tend to read anywhere from 9.8 to 10. Not ideal values but still in the range of gravity.

### Checks
- [x] GRAVITY works

### CODE REVIEW

File: Project/DAQ_System/Program/Src/Sensor/Accelerometer/iaccelerometer.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        virtual void init() = 0;
+        virtual void calibrate() = 0;
+
+        virtual void ComputeAcceleration() = 0;
+        virtual void ComputeMagnetometer() = 0;
+
+
+        virtual float* GetAcceleration() = 0;
+        virtual float* GetMagnetometerData() = 0;
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/Accelerometer/iaccelerometer.hpp]: Remove `ComputerMagnetometer()` and `GetMagnetometerData()`.

Those methods should belong to an `IMagnetometer` abstract interface and is outside the scope of this task.

File: Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    real_acceleration_data_[0] = (raw_acceleration_data_[0]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[1] = (raw_acceleration_data_[1]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[2] = (raw_acceleration_data_[2]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+*/
+
+    // linearization of the accelerometer. by default range of the accelerometer is ±2 g
+    real_acceleration_data_[0] = 0.0006 *raw_acceleration_data_[0] - 0.0002;
+    real_acceleration_data_[1] = 0.0006 *raw_acceleration_data_[1] - 0.0002;
+    real_acceleration_data_[2] = 0.0006 *raw_acceleration_data_[2] - 0.0002;
+}
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp:127]: What are the tradeoffs between your linearized approach vs. the old approach?

I admit, I don't remember the meaning behind `kMG_LSB` — probably just kept from Arduino's library. I recall seeing a value we did not quite understand just yet, nor can I find a reference to this in the datasheet.

That being said, your approach is much more favorable.

File: Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+void LSM303DLHC::ReadRawAcceleration() {
+    uint8_t command[1] = { OUT_X_L_A | 0x80 };
+    HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, command,1, HAL_MAX_DELAY);
+
+    uint8_t bytes_received[6];
+    HAL_I2C_Master_Receive(&i2c_,ACC_ADDRESS, bytes_received,6, HAL_MAX_DELAY);
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp]: Add constants for the amount of bytes to send and receive — for the sake of less "magic numbers".

**Review by crsz20 — COMMENTED** [review]
Pretty solid changes!

File: Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp (hunk 2 above)
Comment (WilliamLim781) [reply Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp:127]: I didn't understand the original approach, and neither could I find any information on it. So when I did research on optimizing accelerometers, Linearizing the data was the most approachable/easiest way I found to do the accelerometer. 

When it comes to tradeoffs, the first thing that comes to mind is that we need to have experimental data to use for our reference points for the linearization. To find the 1 g values for the accelerometer is easy, but to find experimental data for other reference points could be tricky and would require some good testing methodology to achieve( I.E how would we reliably test 1.5 g or other values). 

Another potential tradeoff is that assuming the accelerometer is completely linear from -2g to 2g could be a factually wrong assumption. This is something that, admittedly, I cannot confirm or deny this claim but it is some food for thought when we can further test the accelerometer's accuracy.

**Review by dani4ko03 — APPROVED** [review]
Nice work

**Review by crsz20 — APPROVED** [review]

**Review by crsz20 — APPROVED** [review]
